### PR TITLE
feat: upload to candy machine

### PR DESCRIPTION
**Pull request by thlorenz**
## Summary

This PR adds the preliminary API to upload assets to the candy machine.

- `uploadAssetForCandyMachine` allows to upload a single asset and optionally add it to the
  candy machine
- `uploadAssetsForCandyMachine` allows to upload multiple assets and optionally add them to the
  candy machine
  - it is configurable whether those uploads happen in parallel or sequentially
- the `CandyMachineClient` has been split into multiple modules in order to better organize
  parameter types toghether with the methods they are used with
- the one module for the client was growing large to the point of getting unmanagable
- _internal_ Candy Machine helpers to count config lines + get the assets stored in them were
  added and are exposed via the CandyMachine model

## Related Work

- token-metadata program upgraded
- solana/web3.js version upgraded to latest minor
- transaction failed error now includes logs in message
- use amman-client + latest amman
  
### Input Params

I'm not too happy with the inputs to the upload methods.
Similar the inputs for the NFT plugin methods aren't always ideal.

As discussed with @lorisleiva we will overhaul those at a later time.

### Web3.js Deprecated API

Currently using `TransactionCtorFields_DEPRECATED` which we should change to `TransactionBlockhashCtor`.

FIXES: #109
FIXES: #52

### CODE REVIEW

File: test/plugins/candyMachineModule/uploadAssets.test.ts [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  }
+}
+
+function setupMockStorage(mx: Metaplex) {
+  if (process.env.CI == null) {
+    mx.use(ammanMockStorage(MOCK_STORAGE_ID));
+  } else {
+    mx.use(mockStorage());
+  }
+}
```
Comment (lorisleiva) [inline test/plugins/candyMachineModule/uploadAssets.test.ts:76]: Should we set that up in the global `metaplex` helper function instead?

**Review by lorisleiva — CHANGES_REQUESTED** [review]
Just a few comments and suggestions but I really like that. 🍺

File: src/programs/candyMachine/accounts/candyMachineInternals.ts [hunk 2]
```diff
@@ -59,3 +64,56 @@ export function getSpaceForCandy(data: CandyMachineData) {
...
+  }
+}
+
+export function assertCreators(creators: Creator[]) {
+  if (creators.length > MAX_CREATOR_LIMIT) {
+    throw new Error(
+      `Candy Machine creators too long: ${creators} (max ${MAX_CREATOR_LIMIT})`
+    );
+  }
+}
```
[the remark this answers is not in the document]
Comment (thlorenz) [reply src/programs/candyMachine/accounts/candyMachineInternals.ts:119]: Those only get raised when the user calls the appropriate function on the model, i.e. https://github.com/metaplex-foundation/js-next/blob/thlorenz/feat/upload/src/programs/candyMachine/accounts/CandyMachineAccount.ts#L30

However when processing operations they are wrapped in a candy machine error: https://github.com/metaplex-foundation/js-next/blob/thlorenz/feat/upload/src/plugins/candyMachineModule/Client.helpers.ts#L44-L54

File: test/plugins/candyMachineModule/addAssets.test.ts [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+  // When I add one asset with a uri that is longer than 200 characters
+  try {
+    t.comment('Adding one asset with a uri that is longer than 200 characters');
+    await cm.addAssets({
+      authoritySigner: payerSigner,
+      candyMachineAddress: candyMachineSigner.publicKey,
+      assets: [
+        {
+          name: 'asset',
+          uri: 'uri - 0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|0123456789|',
```
[the remark this answers is not in the document]
Comment (thlorenz) [reply test/plugins/candyMachineModule/addAssets.test.ts:359]: I wanted to make clear in the test that this is a long string, but I updated.

File: test/plugins/candyMachineModule/uploadAssets.test.ts (hunk 1 above)
Comment (thlorenz) [reply test/plugins/candyMachineModule/uploadAssets.test.ts:76]: I don't think so as this makes it clear that only for this test do we need mock storage.

**Comment by steveluscher** [comment]
> Currently using TransactionCtorFields_DEPRECATED which we should change to TransactionBlockhashCtor.

Thank you!

In the meantime, whatever you do, don't import that type explicitly. We never should have exported it.

Instead, _infer_ it: `ConstructorParameters<typeof Transaction>[0]`.

See: https://github.com/metaplex-foundation/js/pull/205

File: src/programs/candyMachine/accounts/candyMachineInternals.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (thlorenz) [reply src/programs/candyMachine/accounts/candyMachineInternals.ts:119]: OK will address this.

**Review by lorisleiva — APPROVED** [review]

**Pull request merged**